ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT/ChainState.hs

```diff
@@ -281,26 +288,69 @@ append k n signer@(PBftSigner _ gk) PBftChainState{..} = assertInvariant k n $
       , preWindow  = preWindow'
       , inWindow   = inWindow'
       , counts     = updateCounts counts
+      , ebbs       = pruneEBBsLTE (pbftSignerSlotNo <$> mbLatestDropped) ebbs
       }
   where
-    (preAnchor', postAnchor') =
+    (mbLatestDropped, preAnchor', postAnchor') =
       case postAnchor of
         x :<| xs | size postAnchor == maxRollbacks k ->
           prune k n (preAnchor |> x, xs |> signer)
         _otherwise ->
-          (preAnchor, postAnchor |> signer) -- We assume k >= 1
+          (Nothing, preAnchor, postAnchor |> signer) -- We assume k >= 1
 
-    ((preWindow', inWindow'), updateCounts) =
+    ((_, preWindow', inWindow'), updateCounts) =
       case inWindow of
         x :<| xs | size inWindow == getWindowSize n ->
           ( prune k n (preWindow |> x, xs |> signer)
           , incrementKey gk . decrementKey (pbftSignerGenesisKey x)
           )
         _otherwise ->
-          ( (preWindow, inWindow |> signer)
+          ( (Nothing, preWindow, inWindow |> signer)
           , incrementKey gk
           )
 
+pruneEBBsLTE ::
+     Maybe SlotNo
+  -> Map SlotNo (WithOrigin SlotNo) -> Map SlotNo (WithOrigin SlotNo)
+pruneEBBsLTE Nothing  = id
+pruneEBBsLTE (Just s) = snd . Map.split s
+
+pruneEBBsGT ::
+     WithOrigin SlotNo
+  -> Map SlotNo (WithOrigin SlotNo) -> Map SlotNo (WithOrigin SlotNo)
+pruneEBBsGT Origin = const Map.empty
+pruneEBBsGT (At s) = fst . Map.split (succ s)
+
+-- | Append an EBB
+appendEBB :: forall c. (PBftCrypto c, HasCallStack)
+          => SecurityParam
+          -> WindowSize
+          -> SlotNo
+          -> PBftChainState c -> PBftChainState c
+appendEBB k n s PBftChainState{..} = assertInvariant k n $
+    PBftChainState {
+        preAnchor  = preAnchor
+      , postAnchor = postAnchor
+      , preWindow  = preWindow
+      , inWindow   = inWindow
+      , counts     = counts
+      , ebbs       = Map.insert s latestNonEBBSlot ebbs
+      }
+  where
+    latestNonEBBSlot =
+      case inWindow of
+        Empty                                       -> Origin
+        _ :|> x | s >= s'                           -> err
```

Comment:
`s` = slot of the EBB, `s'` = slot of the last signer in the window. Shouldn't this condition be: `s' > s`? I think it would be more readable to rename `s` and `s'` to `ebbSlot` and `lastSignerSlot`.

Reply:
I renamed the variables and made the logic more direct.